### PR TITLE
[RFC] treewide: some desktop file cleanups

### PR DESCRIPTION
**Pull request by K900**
###### Motivation for this change
Started off trying to fix desktop files for nixpkgs-mozilla Firefoxen, now we here.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Firefox changes are looking good to me.

**Comment by K900** [comment]
Ran nixpkgs-review, things built fine.

File: pkgs/applications/version-management/smartgithg/default.nix [hunk 1]
```diff
@@ -76,12 +76,10 @@ stdenv.mkDerivation rec {
...
     startupNotify = "true";
-    extraEntries = ''
-      Keywords=git
-      StartupWMClass=${name}
-      Version=1.0
-      Encoding=UTF-8
-    '';
+    startupWMClass = name;
+    extraDesktopEntries = {
+      Keywords= "git";
```
Comment (sersorrel) [inline pkgs/applications/version-management/smartgithg/default.nix:81]:
```suggestion
      Keywords = "git";
```

Comment (K900) [reply pkgs/applications/version-management/smartgithg/default.nix]: Thanks, applied.

**Comment by K900** [comment]
image

File: pkgs/build-support/make-desktopitem/default.nix [hunk 2]
```diff
@@ -1,67 +1,96 @@
...
+    # "SingleMainWindow" = boolOrNullToString singleMainWindow;
+  } // extraConfig;
+
+  renderLine = name: value: if value != null then "${name}=${value}" else null;
+
+  renderSection = sectionName: attrs:
+    let
+      lines = builtins.filter (v: v != null) (lib.mapAttrsToList renderLine attrs);
+    in
+      builtins.concatStringsSep "\n" ([ "[${sectionName}]" ] ++ lines);
```
Comment (danth) [inline pkgs/build-support/make-desktopitem/default.nix:77]: IMO use of `pipe` here would be easier to understand:
```suggestion
    lib.pipe attrs [
      (lib.mapAttrsToList renderLine)
      (builtins.filter (v: !isNull v))
      (builtins.concatStringsSep "\n")
      (section: ''
        [${sectionName}]
        ${section}
      '')
    ];
```

**Review by danth — APPROVED** [review]
Looked through all the changes, nothing seems out of place to me.

File: pkgs/build-support/make-desktopitem/default.nix (hunk 2 above)
Comment (K900) [reply pkgs/build-support/make-desktopitem/default.nix:77]: Thanks, I do like this more. Applied.

**Review by jonringer — APPROVED** [review]
did a very quick scan, but everything looks good.

**Comment by K900** [comment]
Pushed a small fix to make desktop files with actions generate correctly again.

**Review by piegamesde — COMMENTED** [review]
Change generally looks good, but I have a few nits:

- This does not appear to be backwards compatible, so it should be posted in https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574
- Please add a few code comments explaining the most critical sections, what is happening and why.
- Same thing would be cool for the commit message as well

**Comment by K900** [comment]
Added some comments and wrote up a release note.

**Comment by jonringer** [comment]
I can see this getting stale pretty quickly. are we good to merge?

**Comment by K900** [comment]
I believe so, but let me run one last rebase + eval.

**Comment by mweinelt** [comment]
```
these 2 derivations will be built:
  /nix/store/1iiakhv3l2cchhbf9z6s64jpqm5wad1h-PrusaSlicer-G-code-Viewer.desktop.drv
  /nix/store/b53py1ghfr4infy3w74lckcqqcsqk0rm-prusa-slicer-2.4.0.drv
building '/nix/store/1iiakhv3l2cchhbf9z6s64jpqm5wad1h-PrusaSlicer-G-code-Viewer.desktop.drv'...
/build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 1: G-code: command not found
error: builder for '/nix/store/1iiakhv3l2cchhbf9z6s64jpqm5wad1h-PrusaSlicer-G-code-Viewer.desktop.drv' failed with exit code 127
error: 1 dependencies of derivation '/nix/store/b53py1ghfr4infy3w74lckcqqcsqk0rm-prusa-slicer-2.4.0.drv' failed to build
```

I wonder if this change somehow broke `prusa-slicer`.

**Comment by K900** [comment]
I think I know what happened, I'll prepare a fix shortly.

**Comment by sternenseemann** [comment]
This broke `freeoffice`.

**Comment by K900** [comment]
I did evaluate it, but it looks like something went wrong during the final rebase. Fix is at https://github.com/NixOS/nixpkgs/pull/162488

**Comment by oxalica** [comment]
I noticed that there are still many manual `cp` or `ln` used with `makeDesktopItem`, while we already have the `copyDesktopItems` hook. Should we do another clean up?

**Comment by K900** [comment]
Should be a good idea, but I don't think I have the brain space to go through all 300 of these again :)

**Pull request merged**